File: core/src/main/java/com/vzome/core/edits/HeptagonSubdivision.java

package com.vzome.core.edits;


import com.vzome.core.algebra.AlgebraicField;
import com.vzome.core.algebra.AlgebraicNumber;
import com.vzome.core.algebra.AlgebraicVector;
import com.vzome.core.commands.Command;
import com.vzome.core.construction.FreePoint;
import com.vzome.core.construction.Point;
import com.vzome.core.construction.Segment;
import com.vzome.core.construction.SegmentJoiningPoints;
import com.vzome.core.editor.api.ChangeManifestations;
import com.vzome.core.editor.api.EditorModel;
import com.vzome.core.model.Connector;
import com.vzome.core.model.Manifestation;

public class HeptagonSubdivision extends ChangeManifestations
{
    @Override
    public void perform() throws Command.Failure
    {
        Point p1 = null;
        setOrderedSelection( true );
        for (Manifestation man : mSelection) {
            unselect( man );
            if ( man instanceof Connector ) {
                Point nextPoint = (Point) man.getFirstConstruction();
                if ( p1 == null )
                    p1 = nextPoint;
                else {
                    Segment segment = new SegmentJoiningPoints( p1, nextPoint );
                    AlgebraicField field = segment .getField();
                    AlgebraicNumber scaleFactor = field .getUnitTerm(2) .reciprocal();
                    AlgebraicVector offset = segment .getOffset();
                    AlgebraicVector off2 = offset .scale( scaleFactor );
                    AlgebraicVector off1 = off2 .scale( scaleFactor );

                    AlgebraicVector v1 = p1.getLocation() .plus( off1 );
                    Point firstPoint = new FreePoint( v1 );
                    select( manifestConstruction( firstPoint ) );

                    AlgebraicVector v2 = v1 .plus( off2 ); // note, starting from v1
                    // not p1
                    Point secondPoint = new FreePoint( v2 );
                    select( manifestConstruction( secondPoint ) );
                    break;
                }
            }
        }

        redo();
    }

    public HeptagonSubdivision( EditorModel editor )
    {
        super( editor );
    }

    @Override
    protected String getXmlElementName()
    {
        return "HeptagonSubdivision";
    }
}

File: online/src/worker/legacy/from-java/com/vzome/core/edits/HeptagonSubdivision.ts
import { AlgebraicField } from "../algebra/AlgebraicField.js";
import { AlgebraicNumber } from "../algebra/AlgebraicNumber.js";
import { AlgebraicVector } from "../algebra/AlgebraicVector.js";
import { FreePoint } from "../construction/FreePoint.js";
import { Point } from "../construction/Point.js";
import { Segment } from "../construction/Segment.js";
import { SegmentJoiningPoints } from "../construction/SegmentJoiningPoints.js";
import { ChangeManifestations } from "../editor/api/ChangeManifestations.js";
import { EditorModel } from "../editor/api/EditorModel.js";

export class HeptagonSubdivision extends ChangeManifestations {
    /**
     * 
     */
    public perform() {
        let p1: Point = null;
        this.setOrderedSelection(true);
        for(let index=this.mSelection.iterator();index.hasNext();) {
            let man = index.next();
            {
                this.unselect$com_vzome_core_model_Manifestation(man);
                if (man != null && (man.constructor != null && man.constructor["__interfaces"] != null && man.constructor["__interfaces"].indexOf("com.vzome.core.model.Connector") >= 0)){
                    const nextPoint: Point = <Point>man.getFirstConstruction();
                    if (p1 == null)p1 = nextPoint; else {
                        const segment: Segment = new SegmentJoiningPoints(p1, nextPoint);
                        const field: AlgebraicField = segment.getField();
                        const scaleFactor: AlgebraicNumber = field.getUnitTerm(2).reciprocal();
                        const offset: AlgebraicVector = segment.getOffset();
                        const off2: AlgebraicVector = offset.scale(scaleFactor);
                        const off1: AlgebraicVector = off2.scale(scaleFactor);
                        const v1: AlgebraicVector = p1.getLocation().plus(off1);
                        const firstPoint: Point = new FreePoint(v1);
                        this.select$com_vzome_core_model_Manifestation(this.manifestConstruction(firstPoint));
                        const v2: AlgebraicVector = v1.plus(off2);
                        const secondPoint: Point = new FreePoint(v2);
                        this.select$com_vzome_core_model_Manifestation(this.manifestConstruction(secondPoint));
                        break;
                    }
                }
            }
        }
        this.redo();
    }

    public constructor(editor: EditorModel) {
        super(editor);
    }

    /**
     * 
     * @return {string}
     */
    getXmlElementName(): string {
        return "HeptagonSubdivision";
    }
}
HeptagonSubdivision["__class"] = "com.vzome.core.edits.HeptagonSubdivision";
